### PR TITLE
Fail OSD deployment early in case of error

### LINKED ISSUE #984: set_bytes exception
With a misconfigured cluster (i.e. drives configured in the pillar that do not exist on the minion), the following error

Exception: '<' not supported between instances of 'NoneType' and 'int'

will be the result.  Not terribly helpful.  

The simple fix is adding a 'return 0'

```
            for disk in disks[__grains__['id']]:
                log.debug("Device File: {}".format(disk['Device File']))
                log.debug("device: {}".format(self.device))
                if disk['Device File'] == self.device:
                    return int(disk['Bytes'])
            return 0
```

This may just run into another error though.  We should skip these gracefully and log that they do not exist.

### LINKED ISSUE #1076: In Stage 3, error in "ceph-disk activate" passes unnoticed
### Description of Issue/Question

When re-deploying a cluster with unzapped bluestore OSD disks, Stage 3 fails with an obscure error ("cephprocesses.wait"). Upon closer investigation, it is found that "ceph-disk activate" returned status code 1 on all disks.

It would be much nicer if the error message reflected the real cause of the failure. I'm marking it as a bug because the error is reported incorrectly.

### Setup

DeepSea master branch

### PR DESCRIPTION
**Pull request by jschmid1**
Signed-off-by: Joshua Schmid <jschmid@suse.de>

Fixes #1076 
Fixes #984 

Description:

If a OSD fails to deploy you will not get past stage.3 as the restart logic will stop you. This patch stops the deployment earlier and therefore saves the user time ($timeout is in the restarts is omitted)

Example error message:

```

                        ID: deploy OSDs
                  Function: module.run
                      Name: osd.deploy
                    Result: False
                   Comment: Module function osd.deploy executed
                   Started: 17:12:07.961774
                  Duration: 3284.045 ms
                   Changes:   
                            ----------
                            ret:
                                ----------
                                msg:
                                    One or more OSDs failed to deploy. Please see the minion logs
                                result:
                                    False

              Summary for data4.ceph
              ------------
              Succeeded: 2 (changed=1)
              Failed:    1
              ------------
              Total states run:     3
              Total run time:   3.349 s
```

-----------------

[Please find more information on how to run integration tests here](https://github.com/SUSE/DeepSea/wiki/Integration-tests)

### CODE REVIEW

File: srv/salt/_modules/osd.py [hunk 1]
```diff
@@ -1884,15 +1885,36 @@ def deploy():
...
+                       Stdout: {}
+                       Stderr: {}
+                       """.format(func_name,
+                                  device,
+                                  run_data[0],
+                                  run_data[1],
+                                  run_data[2]))
+                log.error(msg)
+    if failed_run:
+        return {'msg': 'One or more OSDs failed to deploy. Please see the minion logs',
```
Comment (jschmid1) [inline srv/salt/_modules/osd.py]: I probably could be precise in the counting :>

**Comment by swiftgist** [comment]
If I have 24 drives and the second one fails, does this not force the admin to deal with the second drive before continuing?  In the current behavior, all could succeed except the second drive.

**Comment by jschmid1** [comment]
> If I have 24 drives and the second one fails, does this not force the admin to deal with the second drive before continuing? In the current behavior, all could succeed except the second drive.

Yes, that's on purpose. I think it's more convenient for the user to get all the drives configured instead of being stopped at 2/24, fixing the error and running the stages over and over again. 
With this 'error-collecting' approach we basically do what we can and then bail out if an error occured.

**Comment by Martin-Weiss** [comment]
> I think it's more convenient for the user to get all the drives configured instead of being stopped at 2/24,

I believe this really depends... 

In small clusters where it is unlikely to have "dead on arrival" disks this is the right approach IMO, too - especially due to issue #1093 where during each re-run of stage.3 unused and orphaned partitions will be created that need to be removed manually..

However - in large clusters - think of 100 nodes with 2400 OSDs - and where we have maybe 50 disks "dead on arrival" - stopping and restarting stage.3 50 times might not be the best approach.

**Comment by swiftgist** [comment]
I think Martin's comment is fair and that "it depends".  We don't have trivial way to allow an admin to pick one strategy or the other.  I think the bigger issue is we still don't have a nice way of presenting failures back up the stack.  I think some of this starts to hit on how Salt works and that we have "terse" as the default setting.  And not to get carried away, but allowing one site to use both strategies for different minions would be the most flexible.  (Think of one troublesome server where any working is okay vs. the rest of the cluster that the admin wants strict failures.)

**Comment by jschmid1** [comment]
ok, let's close this for now. We can still revive it later.

**Pull request closed without merging**